### ISSUE TITLE
Calckey/Misskey API support

### ISSUE THREAD

**Issue by ThatOneCalculator** [GeopJr/Tooth#32]
Having this act as more of a cross-fediverse client rather than specifically a Mastodon client would help differentiate it :)

**Comment by GeopJr** [comment]
I really like Misskey/Calckey so I'm definitely interested in supporting them!

I gave it some tries since this issue opened but it's semi-blocked by #12 and #14. In the meantime if/when Calckey gets Mastodon API compat, similar to Pleroma or Akkoma, it should be supported by Tooth!

**Comment by McSinyx** [comment]
Would (custom) emoji reactions be supported?  This also affects Pleroma and Akkoma, and it requires a layout change.

**Comment by ThatOneCalculator** [comment]
@GeopJr I don't think they'll ever get Masto API support to be honest, it'd moreso be you implementing the Mk API

**Comment by GeopJr** [comment]
> Would (custom) emoji reactions be supported? This also affects Pleroma and Akkoma, and it requires a layout change.

Opened another issue for it #44 

> @GeopJr I don't think they'll ever get Masto API support to be honest, it'd moreso be you implementing the Mk API

I mostly went by this comment: https://codeberg.org/calckey/calckey/issues/9256#issuecomment-731083 but since the issue is now closed I take it as not planned. (Either way, when I said "in the meantime" I meant until those 2 issues get fixed and full mk/ck API support gets added!)

**Comment by ThatOneCalculator** [comment]
@GeopJr we now support the Mastodon API! You can test it out on https://stop.voring.me

It works sort of, but there's an issue of no images loading.

![image](https://user-images.githubusercontent.com/44733677/227388804-d6cf23d8-a0d5-4663-9526-2c752ebeee99.png)

**Comment by GeopJr** [comment]
I'm having trouble reproducing, could you send me your debugging information? (About > Troubleshooting > Debugging Information)

Also, could you are give running from a terminal a try? There will be a lot of logs so keep an eye around `Network.vala:50: GET: https://calckey.social/files/36d58ab7-6931-47f3-a5a5-e21889fe1ce7` for media urls

![image](https://user-images.githubusercontent.com/18014039/227503316-1035e3ee-e710-4c4c-9955-bf5b69cdf0cb.png)

**Comment by GeopJr** [comment]
On further debugging it, I found the following issues:

- When selecting a profile it 401s (`/api/v1/accounts/$id/statuses`) `Request "https://calckey.social/api/v1/accounts/9arjouq4ja/statuses?pinned=true" failed: 401 Unauthorized`
- Account header was null, probably related to the above

Selecting my own profile however, seems to use the Mastodon id instead `https://calckey.social/api/v1/accounts/057099112122053114116103113105/statuses?only_media=false&exclude_replies=true&limit=20` and a http.cat header is available

**Comment by ThatOneCalculator** [comment]
Interesting, I'll investigate that.

Can you also make an account on https://stop.voring.me? calckey.social stores its files on disk, but stop.voring.me uses a CDN, and I get this error:

```
(dev.geopjr.Tuba:16293): Tuba-WARNING **: 12:01:58.253: ImageCache.vala:44: Failed to download image at "https://pool.jortage.com/voringme/misskey/thumbnail-0b704e35-8d0d-4976-9c32-c4e3d8457007.webp". Unrecognized image file format.
```

**Comment by GeopJr** [comment]
Still can't reproduce, my only speculation is that you are missing webp libraries, I think it's `webp-pixbuf-loader`, could you give it a try?

**Comment by ThatOneCalculator** [comment]
Yep, that fixed it. That should probably be added as a dependency for the AUR.

**Comment by GeopJr** [comment]
cc: @fabiscafe

**Comment by ThatOneCalculator** [comment]
One more bug: unicode emoji reactions show, but custom emoji reactions don't.

![image](https://user-images.githubusercontent.com/44733677/227619728-f82a8cfd-6a07-4d36-af04-08205b638b1d.png)

**Comment by GeopJr** [comment]
Pleroma and Mastodon forks (that support reactions), provide the url for custom emojis:
```
{
      "count" : int,
      "me" : bool,
      "name" : string,
      "url": string?
}
```

Calckey doesn't seem to do so:
https://stop.voring.me/api/v1/statuses/9cn6nnxz8g

```
{
      "count": 1,
      "me": false,
      "name": ":rage1@.:"
}
```

**Comment by fabiscafe** [comment]
> Still can't reproduce, my only speculation is that you are missing webp libraries, I think it's `webp-pixbuf-loader`, could you give it a try?

Update is available :partying_face: 

Would you add a meson check  for it as runtime dependency to the next version? Usually in situations where it is not a compile time dependency I'd add it as optional dependency, but since this is UX breaking it should be a hard dependency (and is on the PKGBUILD)

**Comment by GeopJr** [comment]
Thanks!

I feel like you'll be asked to make it optional since technically it works without it, but if you are not until the next release, I'll add a meson check!

**Comment by fabiscafe** [comment]
Thanks! Let's not tell anyone, so nobody will notice :smirk:

**Comment by VarLad** [comment]
@GeopJr So does Misskey work on Tuba yet? If not, any plans in the future?

**Comment by ThatOneCalculator** [comment]
The compatibility layer is very barebones, so proper API support in the future would be much appreciated.

**Comment by GeopJr** [comment]
Started initial support on this branch: https://github.com/GeopJr/Tuba/tree/experiment/wip/feat/mk-api-support

It's very far from being usable, I've only completed the login flow and basic compat for some objects. Priority is still low. The plan is to first get the tuba 100% usable for Misskey (replacing Mastodon support) and then split it into its own thing so if anyone is planning on testing/contributing it might mess up your saved (and you might have to remove them manually from Passwords & Keys (seahorse).
